messaging/broker.go

```diff
@@ -332,67 +348,249 @@ func (b *Broker) applyUnsubscribe(m *Message) {
-		// TODO: Save index.
-		return nil
-	}
+	// Save highest applied index.
+	// TODO: Persist to disk for raft commands.
+	b.index = e.Index
 
-	// Decode the message from the raft log.
+	// Create a message with the same index as Raft.
 	m := &Message{}
-	err := m.UnmarshalBinary(e.Data)
-	assert(err == nil, "message unmarshal: %s", err)
 
-	// Add the raft index to the message.
-	m.Index = e.Index
-
-	// Update the broker configuration.
-	switch m.Type {
-	case CreateReplicaMessageType:
-		b.applyCreateReplica(m)
-	case DeleteReplicaMessageType:
-		b.applyDeleteReplica(m)
-	case SubscribeMessageType:
-		b.applySubscribe(m)
-	case UnsubscribeMessageType:
-		b.applyUnsubscribe(m)
+	// Decode commands into messages.
+	// Convert internal raft entries to no-ops to move the index forward.
+	if e.Type == raft.LogEntryCommand {
+		// Decode the message from the raft log.
+		err := m.UnmarshalBinary(e.Data)
+		assert(err == nil, "message unmarshal: %s", err)
+
+		// Update the broker configuration.
+		switch m.Type {
+		case CreateReplicaMessageType:
+			b.mustApplyCreateReplica(m)
+		case DeleteReplicaMessageType:
+			b.mustApplyDeleteReplica(m)
+		case SubscribeMessageType:
+			b.mustApplySubscribe(m)
+		case UnsubscribeMessageType:
+			b.mustApplyUnsubscribe(m)
+		}
+	} else {
+		// Internal raft commands should be broadcast out as no-ops.
+		m.TopicID = BroadcastTopicID
+		m.Type = InternalMessageType
 	}
 
+	// Set the raft index.
+	m.Index = e.Index
+
 	// Write to the topic.
 	t := b.createTopicIfNotExists(m.TopicID)
 	if err := t.encode(m); err != nil {
-		return fmt.Errorf("encode: %s", err)
+		panic("encode: " + err.Error())
 	}
-
-	return nil
 }
 
 // Index returns the highest index that the broker has seen.
 func (fsm *brokerFSM) Index() (uint64, error) {
-	// TODO: Retrieve index.
-	return 0, nil
+	b := (*Broker)(fsm)
+	return b.index, nil
 }
 
 // Snapshot streams the current state of the broker and returns the index.
 func (fsm *brokerFSM) Snapshot(w io.Writer) (uint64, error) {
+	b := (*Broker)(fsm)
+
 	// TODO: Prevent truncation during snapshot.
-	// TODO: Lock and calculate header.
-	// TODO: Retrieve snapshot index.
-	// TODO: Stream each topic.
-	return 0, nil
+
+	// Calculate header under lock.
+	b.mu.RLock()
+	s, err := fsm.createSnapshot()
+	b.mu.RUnlock()
+	if err != nil {
+		return 0, fmt.Errorf("create snapshot: %s", err)
+	}
+
+	// Encode snapshot header.
+	buf, err := json.Marshal(&s)
+	if err != nil {
+		return 0, fmt.Errorf("encode snapshot header: %s", err)
+	}
+
+	// Write header frame.
+	if err := binary.Write(w, binary.BigEndian, uint32(len(buf))); err != nil {
+		return 0, fmt.Errorf("write header size: %s", err)
+	}
+	if _, err := w.Write(buf); err != nil {
+		return 0, fmt.Errorf("write header: %s", err)
+	}
+
+	// Stream each topic sequentially.
+	for _, t := range s.Topics {
+		if _, err := copyFileN(w, t.path, t.Size); err != nil {
+			return 0, err
+		}
+	}
+
+	// Return the snapshot and its current index.
```

Comment:
The comment says "current index", when it's actually the highest index across all topics. Not sure if this is the *definition* of "current index". I want to confirm this.